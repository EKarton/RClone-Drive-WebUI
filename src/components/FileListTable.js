import PropTypes from 'prop-types';
import FolderIcon from '@mui/icons-material/Folder';
import DescriptionIcon from '@mui/icons-material/Description';
import ImageIcon from '@mui/icons-material/Image';
import Icon from '@mui/material/Icon';
import { Table } from 'semantic-ui-react';
import { ICON_SIZE } from 'utils/constants';
import './FileListTable.scss';

export default function FileListTable({ files, iconSize, onFileClicked }) {
  const renderTableRow = (file) => {
    const handleFileClicked = () => {
      onFileClicked(file);
    };

    return (
      <Table.Row key={file.name}>
        <Table.Cell>
          <div
            className="filelist-table__file-cell"
            onClick={handleFileClicked}
            data-testid={file.name}
          >
            {renderFileIcon(file)}
            <div>{file.name}</div>
          </div>
        </Table.Cell>
        <Table.Cell>{file.lastUpdatedTime}</Table.Cell>
        <Table.Cell>{file.isDirectory ? '-' : file.size}</Table.Cell>
      </Table.Row>
    );
  };

  const renderFileIcon = (file) => {
    if (file.isDirectory) {
      return <FolderIcon fontSize={iconSize} />;
    }

    if (file.isImage) {
      if (file.preview) {
        return (
          <Icon fontSize={iconSize} className="filelist-table__preview-icon">
            {file.preview}
          </Icon>
        );
      }

      return <ImageIcon fontSize={iconSize} />;
    }

    return <DescriptionIcon fontSize={iconSize} />;
  };

  return (
    <Table striped>
      <Table.Header>
        <Table.Row>
          <Table.HeaderCell>Name</Table.HeaderCell>
          <Table.HeaderCell>Date Modified</Table.HeaderCell>
          <Table.HeaderCell>File Size</Table.HeaderCell>
        </Table.Row>
      </Table.Header>
      <Table.Body>{files.map((file) => renderTableRow(file))}</Table.Body>
    </Table>
  );
}

FileListTable.defaultProps = {
  iconSize: ICON_SIZE.LARGE,
};

FileListTable.propType = {
  files: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
      lastUpdatedTime: PropTypes.string,
      size: PropTypes.string,
      isDirectory: PropTypes.bool.isRequired,
      isImage: PropTypes.bool.isRequired,
      preview: PropTypes.node,
    })
  ),
  iconSize: PropTypes.oneOf(Object.values(ICON_SIZE)),
};
