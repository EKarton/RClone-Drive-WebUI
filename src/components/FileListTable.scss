.filelist-table {
  &__file-cell {
    display: flex;
    column-gap: 1rem;
  }
}
