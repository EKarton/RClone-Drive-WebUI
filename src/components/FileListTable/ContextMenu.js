import { Divider, ListItemIcon, ListItemText, Menu, MenuItem } from '@mui/material';
import FileOpenIcon from '@mui/icons-material/FileOpen';
import DownloadIcon from '@mui/icons-material/Download';
import DriveFileMoveIcon from '@mui/icons-material/DriveFileMove';
import ContentCopyIcon from '@mui/icons-material/ContentCopy';
import DriveFileRenameOutlineIcon from '@mui/icons-material/DriveFileRenameOutline';
import DeleteIcon from '@mui/icons-material/Delete';

export default function ContextMenu({
  open,
  onClose,
  menuPosition,
  menuButtonEvents: { onOpen, onDownload, onMove, onCopy, onRename, onDelete },
}) {
  return (
    <Menu
      open={open}
      onClose={onClose}
      anchorReference="anchorPosition"
      anchorPosition={menuPosition}
    >
      <MenuItem onClick={onOpen}>
        <ListItemIcon>
          <FileOpenIcon fontSize="small" />
        </ListItemIcon>
        <ListItemText>Open</ListItemText>
      </MenuItem>
      <MenuItem onClick={onDownload}>
        <ListItemIcon>
          <DownloadIcon fontSize="small" />
        </ListItemIcon>
        <ListItemText>Download</ListItemText>
      </MenuItem>
      <Divider />
      <MenuItem onClick={onMove}>
        <ListItemIcon>
          <DriveFileMoveIcon fontSize="small" />
        </ListItemIcon>
        <ListItemText>Move to</ListItemText>
      </MenuItem>
      <MenuItem onClick={onCopy}>
        <ListItemIcon>
          <ContentCopyIcon fontSize="small" />
        </ListItemIcon>
        <ListItemText>Copy</ListItemText>
      </MenuItem>
      <MenuItem onClick={onRename}>
        <ListItemIcon>
          <DriveFileRenameOutlineIcon fontSize="small" />
        </ListItemIcon>
        <ListItemText>Rename</ListItemText>
      </MenuItem>
      <Divider />
      <MenuItem onClick={onDelete}>
        <ListItemIcon>
          <DeleteIcon fontSize="small" />
        </ListItemIcon>
        <ListItemText>Delete</ListItemText>
      </MenuItem>
    </Menu>
  );
}
