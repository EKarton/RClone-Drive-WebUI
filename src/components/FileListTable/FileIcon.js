import FolderIcon from '@mui/icons-material/Folder';
import DescriptionIcon from '@mui/icons-material/Description';
import ImageIcon from '@mui/icons-material/Image';
import Icon from '@mui/material/Icon';
import './FileIcon.scss';
import LazyImage from 'components/LazyImage';

export default function FileIcon({ file, iconSize }) {
  if (file.isDirectory) {
    return <FolderIcon fontSize={iconSize} />;
  }

  if (file.isImage) {
    return (
      <Icon fontSize={iconSize} className="file-icon__preview-icon">
        <LazyImage
          image={{
            remote: file.remote,
            folderPath: file.path,
            fileName: file.name,
          }}
          imgClassName="file-icon__img"
        />
      </Icon>
    );
  }

  if (file.isImage) {
    return <ImageIcon fontSize={iconSize} />;
  }

  return <DescriptionIcon fontSize={iconSize} />;
}
