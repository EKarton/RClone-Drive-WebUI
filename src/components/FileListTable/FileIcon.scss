.file-icon {
  &__preview-icon {
    display: flex;
    align-items: center;
  }

  &__img {
    width: 100%;
    height: 100%;
    padding: 0 !important;
    object-fit: contain;
  }
}
