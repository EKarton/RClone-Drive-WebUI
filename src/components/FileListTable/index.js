import Paper from '@mui/material/Paper';
import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import PropTypes from 'prop-types';
import StandardRow from '../FileListTableRows/StandardRow/index';
import UploadingRow from '../FileListTableRows/UploadingRow';
import './index.scss';

export default function FileListTable({ children, ...props }) {
  return (
    <TableContainer
      component={Paper}
      className="file-list-table"
      data-testid="file-list-table"
      {...props}
    >
      <Table>
        <TableHead>
          <TableRow>
            <TableCell className="file-list-table__header-cell">Name</TableCell>
            <TableCell className="file-list-table__header-cell">Date Modified</TableCell>
            <TableCell className="file-list-table__header-cell">File Size</TableCell>
          </TableRow>
        </TableHead>
        <TableBody className="file-list-table__body">{children}</TableBody>
      </Table>
    </TableContainer>
  );
}

FileListTable.propTypes = {
  rows: PropTypes.arrayOf(PropTypes.oneOfType([UploadingRow, StandardRow])),
};
