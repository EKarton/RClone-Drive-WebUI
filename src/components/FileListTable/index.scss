.file-list-table {
  margin-bottom: 0 !important;
  box-shadow: none !important;
  border: 1px solid rgba(34 36 38 / 15%);

  &__header-cell {
    background: none !important;
  }
}
