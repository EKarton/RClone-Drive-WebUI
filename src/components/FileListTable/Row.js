import PropTypes from 'prop-types';
import { Table } from 'semantic-ui-react';
import { ICON_SIZE } from 'utils/constants';
import './Row.scss';
import prettyBytes from 'pretty-bytes';
import { useState } from 'react';
import cx from 'classnames';
import ContextMenu from './ContextMenu';
import FileIcon from './FileIcon';

export default function Row({
  file,
  iconSize,
  onFileOpen,
  onFileRename,
  onFileCopy,
  onFileDelete,
  onFileDownload,
  onFileMove,
}) {
  const [isRowFocused, setIsRowFocused] = useState(false);
  const [isContextMenuOpen, setIsContextMenuOpen] = useState(false);
  const [contextMenuPos, setContextMenuPos] = useState(null);

  const handleRowFocus = () => {
    setIsRowFocused(true);
  };

  const handleRowBlurred = () => {
    setIsRowFocused(false);
  };

  const handleDoubleClick = () => {
    onFileOpen();
  };

  const handleContextMenuOpened = (e) => {
    e.preventDefault();

    setContextMenuPos({ left: e.clientX - 2, top: e.clientY - 4 });
    setIsContextMenuOpen(true);
  };

  const handleContextMenuClosed = () => {
    setContextMenuPos(undefined);
    setIsContextMenuOpen(false);
  };

  const handleContextMenuClicked = (func) => () => {
    setContextMenuPos(undefined);
    setIsContextMenuOpen(false);
    func();
  };

  const isRowHighlighted = isRowFocused || isContextMenuOpen;

  return (
    <>
      <Table.Row
        className={cx('filelist-table-row', {
          'filelist-table-row--highlight': isRowHighlighted,
        })}
        onContextMenu={handleContextMenuOpened}
        onFocus={handleRowFocus}
        onBlur={handleRowBlurred}
        onDoubleClick={handleDoubleClick}
        tabIndex="0"
      >
        <Table.Cell className="filelist-table-row__table-cell">
          <div className="filelist-table-row__file-cell" data-testid={file.name}>
            <FileIcon file={file} iconSize={iconSize} />
            <div>{file.name}</div>
          </div>
        </Table.Cell>
        <Table.Cell className="filelist-table-row__table-cell">
          {file.lastUpdatedTime}
        </Table.Cell>
        <Table.Cell className="filelist-table-row__table-cell">
          {file.isDirectory ? '-' : prettyBytes(file.size)}
        </Table.Cell>
      </Table.Row>
      <ContextMenu
        open={isContextMenuOpen}
        onClose={handleContextMenuClosed}
        menuPosition={contextMenuPos}
        menuButtonEvents={{
          onOpen: handleContextMenuClicked(onFileOpen),
          onDownload: handleContextMenuClicked(onFileDownload),
          onMove: handleContextMenuClicked(onFileMove),
          onCopy: handleContextMenuClicked(onFileCopy),
          onRename: handleContextMenuClicked(onFileRename),
          onDelete: handleContextMenuClicked(onFileDelete),
        }}
      />
    </>
  );
}

Row.propType = {
  file: PropTypes.shape({
    name: PropTypes.string,
    lastUpdatedTime: PropTypes.string,
    size: PropTypes.string,
    isDirectory: PropTypes.bool.isRequired,
    isImage: PropTypes.bool.isRequired,
    preview: PropTypes.node,
  }),
  iconSize: PropTypes.oneOf(Object.values(ICON_SIZE)),
};
