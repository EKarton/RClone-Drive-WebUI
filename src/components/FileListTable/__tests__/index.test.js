import StandardRow from 'components/FileListTableRows/StandardRow/index';
import { render } from 'test-utils/react';
import FileListTable from '..';

const files = [
  {
    name: 'Documents',
    lastUpdatedTime: '2021-11-01T21:09:05.237Z',
    size: '',
    isDirectory: true,
    isImage: false,
  },
  {
    name: 'Pictures',
    lastUpdatedTime: '2021-09-28T03:39:11.850Z',
    size: '',
    isDirectory: true,
    isImage: false,
  },
];

describe('FileListTable', () => {
  it('should match snapshot given rows', () => {
    const { baseElement } = render(
      <FileListTable>
        {files.map((file) => (
          <StandardRow file={file} />
        ))}
      </FileListTable>
    );

    expect(baseElement).toMatchSnapshot();
  });
});
