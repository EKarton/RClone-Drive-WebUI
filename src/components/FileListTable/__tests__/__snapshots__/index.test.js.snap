// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FileListTable should match snapshot given rows 1`] = `
<body>
  <div>
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root file-list-table css-11xur9t-MuiPaper-root-MuiTableContainer-root"
      data-testid="file-list-table"
    >
      <table
        class="MuiTable-root css-rqglhn-MuiTable-root"
      >
        <thead
          class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
        >
          <tr
            class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
          >
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium file-list-table__header-cell css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Name
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium file-list-table__header-cell css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              Date Modified
            </th>
            <th
              class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium file-list-table__header-cell css-1ygcj2i-MuiTableCell-root"
              scope="col"
            >
              File Size
            </th>
          </tr>
        </thead>
        <tbody
          class="MuiTableBody-root file-list-table__body css-apqrd9-MuiTableBody-root"
        >
          <tr
            class="MuiTableRow-root filelist-table-base-row css-1q1u3t4-MuiTableRow-root"
            tabindex="0"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-1ex1afd-MuiTableCell-root"
            >
              <div
                class="filelist-table-base-row__file-cell"
                data-testid="Documents"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="FolderIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2h-8l-2-2z"
                  />
                </svg>
                <div>
                  Documents
                </div>
              </div>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-1ex1afd-MuiTableCell-root"
            >
              2021-11-01T21:09:05.237Z
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-1ex1afd-MuiTableCell-root"
            >
              -
            </td>
          </tr>
          <tr
            class="MuiTableRow-root filelist-table-base-row css-1q1u3t4-MuiTableRow-root"
            tabindex="0"
          >
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-1ex1afd-MuiTableCell-root"
            >
              <div
                class="filelist-table-base-row__file-cell"
                data-testid="Pictures"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="FolderIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10 4H4c-1.1 0-1.99.9-1.99 2L2 18c0 1.1.9 2 2 2h16c1.1 0 2-.9 2-2V8c0-1.1-.9-2-2-2h-8l-2-2z"
                  />
                </svg>
                <div>
                  Pictures
                </div>
              </div>
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-1ex1afd-MuiTableCell-root"
            >
              2021-09-28T03:39:11.850Z
            </td>
            <td
              class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium filelist-table-base-row__table-cell css-1ex1afd-MuiTableCell-root"
            >
              -
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
</body>
`;
